python/lsst/pipe/base/butlerQuantumContext.py

```diff
@@ -0,0 +1,204 @@
+
+
+class RefDirection(Enum):
+    IN = auto()
+    OUT = auto()
+
+
+class ButlerQuantumContext:
+    """Butler like class specialized for a single quantum
+
+    A ButlerQuantumContext class wraps a standard butler interface and
+    specializes it to the context of a given quantum. What this means
+    in practice is that the only gets and puts that this class allows
+    are DatasetRefs that are contained in the quantum.
+
+    In the future this class will also be used to record providence on
+    what was actually get and put. This is in contrast to what the
+    preflight expects to be get and put by looking at the graph before
+    execution.
+
+    Parameters
+    ----------
+    butler : `lsst.daf.butler.Butler`
+        Butler object from/to which datasets will be get/put
+    quantum : `lsst.daf.butler.core.Quantum`
+        Quantum object that describes the data-sets which will
+        be get/put by a single execution of this node in the
+        pipeline graph.
+    """
+    def __init__(self, butler: Butler, quantum: Quantum):
+        self.quantum = quantum
+        self.registry = butler.registry
+        self.allInputs = set()
+        self.allOutputs = set()
+        for refs in quantum.predictedInputs.values():
+            for ref in refs:
+                self.allInputs.add((ref.datasetType, ref.dataId))
+        for refs in quantum.outputs.values():
+            for ref in refs:
+                self.allOutputs.add((ref.datasetType, ref.dataId))
+
+        # Create closures over butler to discourage anyone from directly
+        # using a butler reference
+        def _get(self, ref):
+            if isinstance(ref, DeferredDatasetRef):
+                self._checkMembership(ref.datasetRef, RefDirection.IN)
+                return butler.getDeferred(ref.datasetRef)
+
+            else:
+                self._checkMembership(ref, RefDirection.IN)
+                return butler.get(ref)
+
+        def _put(self, value, ref):
+            self._checkMembership(ref, RefDirection.OUT)
+            butler.put(value, ref)
+
+        self._get = types.MethodType(_get, self)
+        self._put = types.MethodType(_put, self)
+
+    def get(self, dataset: typing.Union[InputQuantizedConnection, typing.List[DatasetRef], DatasetRef]):
+        """Fetches data from the butler
+
+        Parameters
+        ----------
+        dataset : `InputQuantizedConnection` or `list` of `lsst.daf.butler.DatasetRef`
+                or `lsst.daf.butler.DatasetRef`
+            This argument may either be an `InputQuantizedConnection` which describes
+            all the inputs of a quantum, a list of `lsst.daf.butler.DatasetRef`, or
+            a single `lsst.daf.butler.DatasetRef`. The function will get and return
+            the corresponding datasets from the butler.
+
+        Raises
+        ------
+        ValueError
+            If a `DatasetRef` is passed to get that is not defined in the quantum object
```

Comment:
Needs module name for `DatasetRef`. I'd add a `~` for each of those too to reduce verbosity in generated docs.